src/k8s/cmd/k8s/k8s_refresh_certs.go

```diff
@@ -0,0 +1,77 @@
+package k8s
+
+import (
+	"context"
+	"time"
+
+	apiv1 "github.com/canonical/k8s-snap-api/api/v1"
+	cmdutil "github.com/canonical/k8s/cmd/util"
+	"github.com/canonical/k8s/pkg/utils"
+	"github.com/spf13/cobra"
+)
+
+func newRefreshCertsCmd(env cmdutil.ExecutionEnvironment) *cobra.Command {
+	var opts struct {
+		extraSANs []string
+		ttl       string
+		timeout   time.Duration
+	}
+	cmd := &cobra.Command{
+		Use:    "refresh-certs",
+		Short:  "Refresh the certificates of the running node",
+		PreRun: chainPreRunHooks(hookRequireRoot(env)),
+		Run: func(cmd *cobra.Command, args []string) {
+			ttl, err := utils.TTLToSeconds(opts.ttl)
+			if err != nil {
+				cmd.PrintErrf("Error: Failed to parse TTL. \n\nThe error was: %v\n", err)
+			}
+
+			client, err := env.Snap.K8sdClient("")
+			if err != nil {
+				cmd.PrintErrf("Error: Failed to create a k8sd client. Make sure that the k8sd service is running.\n\nThe error was: %v\n", err)
+				env.Exit(1)
+				return
+			}
+
+			ctx, cancel := context.WithTimeout(cmd.Context(), opts.timeout)
+			cobra.OnFinalize(cancel)
+
+			plan, err := client.RefreshCertificatesPlan(ctx, apiv1.RefreshCertificatesPlanRequest{})
+			if err != nil {
+				cmd.PrintErrf("Error: Failed to get the certificates refresh plan.\n\nThe error was: %v\n", err)
+				env.Exit(1)
+				return
+			}
+
+			if len(plan.CertificateSigningRequests) > 0 {
+				cmd.Println("The following CertificateSigningRequests should be approved. Run the following commands on any of the control plane nodes of the cluster:")
+				for _, csr := range plan.CertificateSigningRequests {
+					cmd.Printf("k8s kubectl certificate approve %s\n", csr)
+				}
+			}
+
+			runRequest := apiv1.RefreshCertificatesRunRequest{
+				Seed:              plan.Seed,
+				ExpirationSeconds: ttl,
+				ExtraSANs:         opts.extraSANs,
+			}
+
+			cmd.Println("Waiting for certificates to be created...")
+			runResponse, err := client.RefreshCertificatesRun(ctx, runRequest)
+			if err != nil {
+				cmd.PrintErrf("Error: Failed to refresh the certificates.\n\nThe error was: %v\n", err)
+				env.Exit(1)
+				return
+			}
+
+			// TODO: Should we print the expiration time in a human-readable format? Should we return the UNIX epoch in the response?
+			cmd.Printf("Certificates have been successfully refreshed, and will expire at %d.\n", runResponse.ExpirationSeconds)
+		},
+	}
+	cmd.Flags().StringVar(&opts.ttl, "ttl", "", "the time-to-live for the certificates.")
```

Comment:
maybe we can use `DurationVar`? I think we can get rid of `TTLToSeconds` then as well.

Reply:
No, we can't do that. Currently, we're only parsing the TTL according to the constants defined here: https://github.com/canonical/k8s-snap/blob/96634b27d7e15590e2b64ca5867b98a78064926a/src/k8s/pkg/utils/time.go#L10-L20

The proposal enforces the use of `expirationSeconds` as the unit to define the expiration time for new certificates. This approach aligns with the `expirationSeconds` of the `CertificateSigningRequest` resources. Additionally, durations are rendered in nanoseconds in JSON, so `{Expiration: time.Second}` would appear as `{"expiration": 1000000000}`. Using seconds as the unit minimizes the risk of overflow. To address this, we've defined a set of predefined constants (as shown in the code block) that match the exact duration we want our certificates to remain valid.

As far as I know, durations can only be parsed up to hours (we should allow the user to use years, months, days and any other unit parsed by duration), and simply multiplying them could result in potentially missing a few days if we set arbitrary durations for months or years. So instead of stacking these durations by converting back and forth, we should use the predefined constants and delegate the time calculations to the logic in k8sd.

NB: This is similar to the approach used in upstream: https://github.com/kubernetes/kubernetes/blob/fa75c8c7491ed10beff187d60fe60b0db8416ff7/pkg/controller/certificates/signer/signer.go#L213-L231

Reply:
fair point. 
btw, in case someone is interested in why Go does not support >time.Hours:
https://github.com/golang/go/issues/11473

Reply:
Sorry for seeing this late, Thanks a lot @mateoflorido